File: iosApp/UkuleleCompanion/Views/SongwriterModeFlow.swift
import SwiftUI
import shared

/// Guided songwriting flow: choose key → build progression → write lyrics → transpose → save.
struct SongwriterModeFlow: View {
    @StateObject private var songbookViewModel = SongbookViewModel()
    @StateObject private var progressionsViewModel = ProgressionsViewModel()

    @State private var currentStep = 0
    @State private var selectedRoot: Int32 = 0
    @State private var selectedScale: ScaleType = .major
    @State private var chosenDegrees: [ChordDegree] = []
    @State private var songTitle = ""
    @State private var songContent = ""
    @State private var transposeSemitones: Int = 0

    private let totalSteps = 5

    var body: some View {
        VStack(spacing: 0) {
            ProgressView(value: Double(currentStep + 1), total: Double(totalSteps))
                .padding(.horizontal)

            Text("Step \(currentStep + 1) of \(totalSteps)")
                .font(.caption)
                .foregroundStyle(.secondary)
                .padding(.top, 4)
                .accessibilityAddTraits(.updatesFrequently)

            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    stepContent
                }
                .padding()
            }

            Spacer()

            navigationButtons
                .padding()
        }
        .navigationTitle("Start a Song")
        .onChange(of: selectedRoot) { _ in
            chosenDegrees = []
            transposeSemitones = 0
        }
        .onChange(of: selectedScale) { _ in
            chosenDegrees = []
            transposeSemitones = 0
        }
    }

    @ViewBuilder
    private var stepContent: some View {
        switch currentStep {
        case 0: stepChooseKeyScale
        case 1: stepBuildProgression
        case 2: stepAddLyrics
        case 3: stepTranspose
        case 4: stepReviewAndSave
        default: EmptyView()
        }
    }

    // MARK: - Step 1: Choose Key & Scale

    private var stepChooseKeyScale: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Choose a key and scale")
                .font(.title3.bold())
                .accessibilityAddTraits(.isHeader)

            Text("Pick the musical key and scale for your song. This determines which chords will be suggested.")
                .font(.subheadline)
                .foregroundStyle(.secondary)

            Text("Root")
                .font(.headline)

            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 6) {
                    ForEach(0..<12, id: \.self) { pc in
                        let name = Notes.shared.pitchClassToName(pitchClass: Int32(pc))
                        Button(name) { selectedRoot = Int32(pc) }
                            .buttonStyle(.bordered)
                            .tint(selectedRoot == Int32(pc) ? .accentColor : .secondary)
                    }
                }
            }

            Text("Scale")
                .font(.headline)

            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 6) {
                    ForEach(progressionsViewModel.allScaleTypes, id: \.name) { scale in
                        Button(scale.label) { selectedScale = scale }
                            .buttonStyle(.bordered)
                            .tint(selectedScale == scale ? .accentColor : .secondary)
                    }
                }
            }
        }
    }

    // MARK: - Step 2: Build Progression

    private var stepBuildProgression: some View {
        let degrees = progressionsViewModel.diatonicDegreesForScale(selectedScale)

        return VStack(alignment: .leading, spacing: 12) {
            Text("Build a chord progression")
                .font(.title3.bold())
                .accessibilityAddTraits(.isHeader)

            Text("Tap the diatonic chords below to build your progression.")
                .font(.subheadline)
                .foregroundStyle(.secondary)

            Text("Tap chords to add")
                .font(.headline)

            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 6) {
                    ForEach(0..<degrees.count, id: \.self) { i in
                        let degree = degrees[i]
                        let chordRoot = (selectedRoot + degree.interval) % 12
                        let chordName = Notes.shared.enharmonicForKey(
                            pitchClass: Int32(chordRoot),
                            keyRoot: KotlinInt(int: selectedRoot),
                            isMinor: selectedScale == .minor
                        ) + degree.quality
                        Button(chordName) {
                            chosenDegrees.append(degree)
                        }
                        .buttonStyle(.bordered)
                    }
                }
            }

            if !chosenDegrees.isEmpty {
                Text("Your progression (\(chosenDegrees.count) chords)")
                    .font(.headline)

                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 6) {
                        ForEach(0..<chosenDegrees.count, id: \.self) { i in
                            let degree = chosenDegrees[i]
                            let chordRoot = (selectedRoot + degree.interval) % 12
                            let chordName = Notes.shared.enharmonicForKey(
                                pitchClass: Int32(chordRoot),
                                keyRoot: KotlinInt(int: selectedRoot),
                                isMinor: selectedScale == .minor
                            ) + degree.quality
                            Button(chordName) {
                                chosenDegrees.remove(at: i)
                            }
                            .buttonStyle(.borderedProminent)
                        }
                    }
                }

                Button("Reset") { chosenDegrees.removeAll() }
                    .font(.caption)
            } else {
                Text("Tap the chords above to build your progression")
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
            }

            if chosenDegrees.count == 1 {
                Text("Add at least 2 chords")
                    .font(.caption)
                    .foregroundStyle(.red)
            }
        }
    }

    // MARK: - Step 3: Add Lyrics

    private var stepAddLyrics: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Add lyrics and chords")
                .font(.title3.bold())
                .accessibilityAddTraits(.isHeader)

            Text("Write your lyrics below. Tap the chord chips to insert [Chord] markers.")
                .font(.subheadline)
                .foregroundStyle(.secondary)

            TextField("Song title", text: $songTitle)
                .textFieldStyle(.roundedBorder)

            if !chosenDegrees.isEmpty {
                ScrollView(.horizontal, showsIndicators: false) {
                    HStack(spacing: 4) {
                        ForEach(0..<chosenDegrees.count, id: \.self) { i in
                            let degree = chosenDegrees[i]
                            let chordRoot = (selectedRoot + degree.interval) % 12
                            let chordName = Notes.shared.enharmonicForKey(
                                pitchClass: Int32(chordRoot),
                                keyRoot: KotlinInt(int: selectedRoot),
                                isMinor: selectedScale == .minor
                            ) + degree.quality
                            Button(chordName) {
                                songContent += "[\(chordName)]"
                            }
                            .buttonStyle(.bordered)
                            .controlSize(.small)
                            .accessibilityHint("Inserts [\(chordName)] into lyrics")
                        }
                    }
                }
            }

            TextEditor(text: $songContent)
                .frame(minHeight: 200)
                .accessibilityLabel("Song lyrics editor")
                .accessibilityHint("Edit lyrics; chord buttons above insert bracketed chords")
                .overlay(
                    RoundedRectangle(cornerRadius: 8)
                        .stroke(Color(.systemGray4), lineWidth: 1)
                )
        }
    }

    // MARK: - Step 4: Transpose

    private var stepTranspose: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Transpose (optional)")
                .font(.title3.bold())
                .accessibilityAddTraits(.isHeader)

            Text("Shift the key up or down if needed. The transposition will be applied when you save.")
                .font(.subheadline)
                .foregroundStyle(.secondary)

            HStack {
                Spacer()
                Button { transposeSemitones -= 1 } label: {
                    Image(systemName: "minus.circle")
                }
                .accessibilityLabel("Transpose down")

                Text("\(transposeSemitones > 0 ? "+" : "")\(transposeSemitones)")
                    .font(.headline.monospacedDigit())
                    .frame(minWidth: 40)
                    .multilineTextAlignment(.center)

                Button { transposeSemitones += 1 } label: {
                    Image(systemName: "plus.circle")
                }
                .accessibilityLabel("Transpose up")

                if transposeSemitones != 0 {
                    Button("Reset") { transposeSemitones = 0 }
                        .font(.caption)
                }
                Spacer()
            }

            if !songContent.isEmpty && transposeSemitones != 0 {
                Text("Preview:")
                    .font(.headline)

                let transposed = ChordSheetTranspose.shared.transpose(
                    content: songContent,
                    semitones: Int32(transposeSemitones)
                )
                Text(transposed)
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }
        }
    }

    // MARK: - Step 5: Review & Save

    private var stepReviewAndSave: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text("Review and save")
                .font(.title3.bold())
                .accessibilityAddTraits(.isHeader)

            Text("Check your song below, then tap Save to add it to your songbook.")
                .font(.subheadline)
                .foregroundStyle(.secondary)

            let keyName = Notes.shared.pitchClassToName(pitchClass: selectedRoot)
            Text("Key: \(keyName) \(selectedScale.label)")
                .font(.subheadline.bold())
                .foregroundStyle(Color.accentColor)

            if !chosenDegrees.isEmpty {
                let chordNames = chosenDegrees.map { degree -> String in
                    let chordRoot = (selectedRoot + degree.interval) % 12
                    return Notes.shared.enharmonicForKey(
                        pitchClass: Int32(chordRoot),
                        keyRoot: KotlinInt(int: selectedRoot),
                        isMinor: selectedScale == .minor
                    ) + degree.quality
                }
                Text("Progression: \(chordNames.joined(separator: " – "))")
                    .font(.caption)
                    .foregroundStyle(.secondary)
            }

            if transposeSemitones != 0 {
                Text("Transpose: \(transposeSemitones > 0 ? "+" : "")\(transposeSemitones) semitones")
                    .font(.caption)
                    .foregroundStyle(.orange)
            }

            Divider()

            Text(songTitle.isEmpty ? "Untitled" : songTitle)
                .font(.headline)

            let finalContent: String = {
                if transposeSemitones != 0 {
                    return ChordSheetTranspose.shared.transpose(
                        content: songContent,
                        semitones: Int32(transposeSemitones)
                    )
                }
                return songContent
            }()
            Text(finalContent.isEmpty ? "(no lyrics yet)" : finalContent)
                .font(.body)
                .foregroundStyle(.secondary)
        }
    }

    // MARK: - Navigation Buttons

    private var navigationButtons: some View {
        HStack {
            if currentStep > 0 {
                Button("Back") { currentStep -= 1 }
                    .buttonStyle(.bordered)
            }

            Spacer()

            if currentStep < totalSteps - 1 {
                Button("Next") { currentStep += 1 }
                    .buttonStyle(.borderedProminent)
                    .disabled(currentStep == 1 && chosenDegrees.count < 2)
            } else {
                Button("Save") {
                    let finalContent: String
                    if transposeSemitones != 0 {
                        finalContent = ChordSheetTranspose.shared.transpose(
                            content: songContent,
                            semitones: Int32(transposeSemitones)
                        )
                    } else {
                        finalContent = songContent
                    }

                    let finalRoot = (selectedRoot + Int32(transposeSemitones) % 12 + 12) % 12
                    let finalKey = Notes.shared.pitchClassToName(pitchClass: finalRoot) + " " + selectedScale.label

                    let song = StoredSong(
                        id: UUID().uuidString,
                        title: songTitle.isEmpty ? "My Song" : songTitle,
                        artist: "",
                        content: finalContent,
                        key: finalKey,
                        capo: 0,
                        strumPatternName: "",
                        labels: [],
                        createdAt: Date().timeIntervalSince1970 * 1000,
                        updatedAt: Date().timeIntervalSince1970 * 1000
                    )
                    songbookViewModel.save(song: song)

                    if chosenDegrees.count >= 2 {
                        let keyName = Notes.shared.pitchClassToName(pitchClass: selectedRoot)
                        progressionsViewModel.createCustomFromDegrees(
                            name: "\(keyName) \(selectedScale.label) song progression",
                            description: "Created in Songwriter Mode",
                            degrees: chosenDegrees,
                            scaleType: selectedScale
                        )
                    }

                    currentStep = 0
                    songTitle = ""
                    songContent = ""
                    chosenDegrees = []
                    transposeSemitones = 0
                }
                .buttonStyle(.borderedProminent)
                .disabled(songTitle.isEmpty && songContent.isEmpty)
            }
        }
    }
}

File: app/src/main/java/com/baijum/ukufretboard/ui/SongwriterModeFlow.kt
package com.baijum.ukufretboard.ui

import android.widget.Toast
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.FilterChip
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.semantics.heading
import androidx.compose.ui.semantics.liveRegion
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.semantics.LiveRegionMode
import com.baijum.ukufretboard.R
import com.baijum.ukufretboard.data.ChordDegree
import com.baijum.ukufretboard.data.Notes
import com.baijum.ukufretboard.data.Progressions
import com.baijum.ukufretboard.data.ScaleType
import com.baijum.ukufretboard.domain.ChordSheetFormatter
import com.baijum.ukufretboard.domain.ChordSheetTranspose
import com.baijum.ukufretboard.viewmodel.SongbookViewModel

private const val TOTAL_STEPS = 5

/**
 * Guided songwriting flow that walks the user through picking a key/scale,
 * building a progression, writing lyrics, transposing, and sharing.
 */
@Composable
fun SongwriterModeFlow(
    songbookViewModel: SongbookViewModel,
    onSaveProgression: (String, String, List<ChordDegree>, ScaleType) -> Unit,
    modifier: Modifier = Modifier,
) {
    var currentStep by rememberSaveable { mutableIntStateOf(0) }

    var selectedRoot by rememberSaveable { mutableIntStateOf(0) }
    var selectedScale by rememberSaveable { mutableStateOf(ScaleType.MAJOR) }
    var chosenDegrees by rememberSaveable { mutableStateOf(listOf<ChordDegree>()) }
    var songTitle by rememberSaveable { mutableStateOf("") }
    var songContent by rememberSaveable { mutableStateOf("") }
    var transposeSemitones by rememberSaveable { mutableIntStateOf(0) }

    val context = LocalContext.current
    val songSavedMessage = stringResource(R.string.songwriter_song_saved)

    Column(
        modifier = modifier
            .fillMaxSize()
            .padding(16.dp),
    ) {
        Text(
            text = stringResource(R.string.nav_songwriter_mode),
            style = MaterialTheme.typography.headlineSmall,
            fontWeight = FontWeight.Bold,
            modifier = Modifier.semantics { heading() },
        )

        Spacer(modifier = Modifier.height(8.dp))

        LinearProgressIndicator(
            progress = { (currentStep + 1).toFloat() / TOTAL_STEPS },
            modifier = Modifier.fillMaxWidth(),
        )

        val stepLabel = "Step ${currentStep + 1} of $TOTAL_STEPS"
        Text(
            text = stepLabel,
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier
                .padding(top = 4.dp, bottom = 12.dp)
                .semantics { liveRegion = LiveRegionMode.Polite },
        )

        Column(
            modifier = Modifier
                .weight(1f)
                .verticalScroll(rememberScrollState()),
        ) {
            when (currentStep) {
                0 -> StepChooseKeyScale(
                    selectedRoot = selectedRoot,
                    onRootChange = {
                        selectedRoot = it
                        chosenDegrees = emptyList()
                        transposeSemitones = 0
                    },
                    selectedScale = selectedScale,
                    onScaleChange = {
                        selectedScale = it
                        chosenDegrees = emptyList()
                        transposeSemitones = 0
                    },
                )
                1 -> StepBuildProgression(
                    selectedRoot = selectedRoot,
                    selectedScale = selectedScale,
                    chosenDegrees = chosenDegrees,
                    onDegreesChange = { chosenDegrees = it },
                )
                2 -> StepAddLyrics(
                    songTitle = songTitle,
                    onTitleChange = { songTitle = it },
                    songContent = songContent,
                    onContentChange = { songContent = it },
                    selectedRoot = selectedRoot,
                    chosenDegrees = chosenDegrees,
                )
                3 -> StepTranspose(
                    content = songContent,
                    transposeSemitones = transposeSemitones,
                    onTransposeChange = { transposeSemitones = it },
                )
                4 -> StepReviewAndSave(
                    songTitle = songTitle,
                    songContent = songContent,
                    transposeSemitones = transposeSemitones,
                    selectedRoot = selectedRoot,
                    chosenDegrees = chosenDegrees,
                    selectedScale = selectedScale,
                )
            }
        }

        Spacer(modifier = Modifier.height(8.dp))

        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
        ) {
            if (currentStep > 0) {
                OutlinedButton(onClick = { currentStep-- }) {
                    Text(stringResource(R.string.onboarding_back))
                }
            } else {
                Spacer(modifier = Modifier.width(1.dp))
            }

            if (currentStep < TOTAL_STEPS - 1) {
                Button(
                    onClick = { currentStep++ },
                    enabled = when (currentStep) {
                        1 -> chosenDegrees.size >= 2
                        else -> true
                    },
                ) {
                    Text(stringResource(R.string.onboarding_next))
                }
            } else {
                Button(
                    onClick = {
                        val finalContent = if (transposeSemitones != 0) {
                            ChordSheetTranspose.transpose(songContent, transposeSemitones)
                        } else {
                            songContent
                        }
                        val finalRoot = ((selectedRoot + transposeSemitones) % 12 + 12) % 12
                        val finalKey = "${Notes.pitchClassToName(finalRoot)} ${selectedScale.label}"
                        // Always create a NEW song here. Songwriter Mode is never
                        // editing an existing sheet, but saveSheet would infer that
                        // from a stale _currentSheet and overwrite the last-opened
                        // song in place (issue #575).
                        songbookViewModel.createSheet(
                            title = songTitle.ifBlank { "My Song" },
                            artist = "",
                            content = finalContent,
                            key = finalKey,
                        )
                        if (chosenDegrees.size >= 2) {
                            val keyName = Notes.pitchClassToName(selectedRoot)
                            onSaveProgression(
                                "$keyName ${selectedScale.name} song progression",
                                "Created in Songwriter Mode",
                                chosenDegrees,
                                selectedScale,
                            )
                        }
                        Toast.makeText(context, songSavedMessage, Toast.LENGTH_SHORT).show()
                        currentStep = 0
                        songTitle = ""
                        songContent = ""
                        chosenDegrees = emptyList()
                        transposeSemitones = 0
                    },
                    enabled = songTitle.isNotBlank() || songContent.isNotBlank(),
                ) {
                    Text(stringResource(R.string.dialog_save))
                }
            }
        }
    }
}

@Composable
private fun StepChooseKeyScale(
    selectedRoot: Int,
    onRootChange: (Int) -> Unit,
    selectedScale: ScaleType,
    onScaleChange: (ScaleType) -> Unit,
) {
    Text(
        text = "Choose a key and scale",
        style = MaterialTheme.typography.titleMedium,
        fontWeight = FontWeight.Bold,
        modifier = Modifier.semantics { heading() },
    )
    Spacer(modifier = Modifier.height(4.dp))
    Text(
        text = "Pick the musical key and scale for your song. This determines which chords will be suggested.",
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
    Spacer(modifier = Modifier.height(16.dp))

    Text(
        text = stringResource(R.string.label_root),
        style = MaterialTheme.typography.labelMedium,
        fontWeight = FontWeight.Bold,
    )
    Spacer(modifier = Modifier.height(4.dp))
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState()),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        for (pc in 0 until Notes.PITCH_CLASS_COUNT) {
            val name = Notes.pitchClassToName(pc)
            FilterChip(
                selected = pc == selectedRoot,
                onClick = { onRootChange(pc) },
                label = { Text(name) },
            )
        }
    }

    Spacer(modifier = Modifier.height(16.dp))

    Text(
        text = stringResource(R.string.label_scale),
        style = MaterialTheme.typography.labelMedium,
        fontWeight = FontWeight.Bold,
    )
    Spacer(modifier = Modifier.height(4.dp))
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState()),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        ScaleType.entries.forEach { scale ->
            FilterChip(
                selected = scale == selectedScale,
                onClick = { onScaleChange(scale) },
                label = { Text(scale.label) },
            )
        }
    }
}

@Composable
private fun StepBuildProgression(
    selectedRoot: Int,
    selectedScale: ScaleType,
    chosenDegrees: List<ChordDegree>,
    onDegreesChange: (List<ChordDegree>) -> Unit,
) {
    Text(
        text = "Build a chord progression",
        style = MaterialTheme.typography.titleMedium,
        fontWeight = FontWeight.Bold,
        modifier = Modifier.semantics { heading() },
    )
    Spacer(modifier = Modifier.height(4.dp))
    Text(
        text = "Tap the diatonic chords below to build your progression. These are the chords that naturally fit your chosen key.",
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
    Spacer(modifier = Modifier.height(16.dp))

    val availableDegrees = Progressions.diatonicDegrees(selectedScale)

    Text(
        text = stringResource(R.string.progression_tap_chords),
        style = MaterialTheme.typography.labelMedium,
        fontWeight = FontWeight.Bold,
    )
    Spacer(modifier = Modifier.height(4.dp))
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState()),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        availableDegrees.forEach { degree ->
            val chordRoot = (selectedRoot + degree.interval) % Notes.PITCH_CLASS_COUNT
            val chordName = Notes.enharmonicForKey(chordRoot, selectedRoot) + degree.quality
            FilterChip(
                selected = false,
                onClick = { onDegreesChange(chosenDegrees + degree) },
                label = { Text(chordName) },
            )
        }
    }

    Spacer(modifier = Modifier.height(16.dp))

    if (chosenDegrees.isNotEmpty()) {
        Text(
            text = stringResource(R.string.progression_your_chords, chosenDegrees.size),
            style = MaterialTheme.typography.labelMedium,
            fontWeight = FontWeight.Bold,
        )
        Spacer(modifier = Modifier.height(4.dp))
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(6.dp),
        ) {
            chosenDegrees.forEachIndexed { index, degree ->
                val chordRoot = (selectedRoot + degree.interval) % Notes.PITCH_CLASS_COUNT
                val chordName = Notes.enharmonicForKey(chordRoot, selectedRoot) + degree.quality
                FilterChip(
                    selected = true,
                    onClick = {
                        onDegreesChange(chosenDegrees.toMutableList().apply { removeAt(index) })
                    },
                    label = { Text(chordName) },
                )
            }
        }
        Spacer(modifier = Modifier.height(4.dp))
        TextButton(onClick = { onDegreesChange(emptyList()) }) {
            Text(stringResource(R.string.dialog_reset))
        }
    } else {
        Text(
            text = stringResource(R.string.progression_placeholder),
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }

    if (chosenDegrees.size == 1) {
        Spacer(modifier = Modifier.height(4.dp))
        Text(
            text = stringResource(R.string.progression_error_min),
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.error,
        )
    }
}

@Composable
private fun StepAddLyrics(
    songTitle: String,
    onTitleChange: (String) -> Unit,
    songContent: String,
    onContentChange: (String) -> Unit,
    selectedRoot: Int,
    chosenDegrees: List<ChordDegree>,
) {
    Text(
        text = "Add lyrics and chords",
        style = MaterialTheme.typography.titleMedium,
        fontWeight = FontWeight.Bold,
        modifier = Modifier.semantics { heading() },
    )
    Spacer(modifier = Modifier.height(4.dp))
    Text(
        text = "Write your lyrics below. Tap the chord chips to insert [Chord] markers.",
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
    Spacer(modifier = Modifier.height(12.dp))

    OutlinedTextField(
        value = songTitle,
        onValueChange = onTitleChange,
        label = { Text(stringResource(R.string.songbook_field_title)) },
        modifier = Modifier.fillMaxWidth(),
        singleLine = true,
    )

    Spacer(modifier = Modifier.height(8.dp))

    if (chosenDegrees.isNotEmpty()) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .horizontalScroll(rememberScrollState()),
            horizontalArrangement = Arrangement.spacedBy(4.dp),
        ) {
            chosenDegrees.forEach { degree ->
                val chordRoot = (selectedRoot + degree.interval) % Notes.PITCH_CLASS_COUNT
                val chordName = Notes.enharmonicForKey(chordRoot, selectedRoot) + degree.quality
                FilterChip(
                    selected = false,
                    onClick = { onContentChange(songContent + "[$chordName]") },
                    label = { Text(chordName) },
                )
            }
        }
        Spacer(modifier = Modifier.height(8.dp))
    }

    OutlinedTextField(
        value = songContent,
        onValueChange = onContentChange,
        label = { Text(stringResource(R.string.songbook_field_lyrics)) },
        modifier = Modifier
            .fillMaxWidth()
            .height(200.dp),
    )
}

@Composable
private fun StepTranspose(
    content: String,
    transposeSemitones: Int,
    onTransposeChange: (Int) -> Unit,
) {
    Text(
        text = "Transpose (optional)",
        style = MaterialTheme.typography.titleMedium,
        fontWeight = FontWeight.Bold,
        modifier = Modifier.semantics { heading() },
    )
    Spacer(modifier = Modifier.height(4.dp))
    Text(
        text = "Shift the key up or down if needed. The transposition will be applied when you save.",
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
    Spacer(modifier = Modifier.height(16.dp))

    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically,
    ) {
        OutlinedButton(onClick = { onTransposeChange(transposeSemitones - 1) }) {
            Text("\u2212")
        }
        Text(
            text = ChordSheetTranspose.semitoneLabel(transposeSemitones),
            style = MaterialTheme.typography.bodyMedium,
            fontWeight = FontWeight.Bold,
            modifier = Modifier.padding(horizontal = 16.dp),
        )
        OutlinedButton(onClick = { onTransposeChange(transposeSemitones + 1) }) {
            Text("+")
        }
        if (transposeSemitones != 0) {
            Spacer(modifier = Modifier.width(8.dp))
            TextButton(onClick = { onTransposeChange(0) }) {
                Text(stringResource(R.string.dialog_reset))
            }
        }
    }

    if (content.isNotEmpty() && transposeSemitones != 0) {
        Spacer(modifier = Modifier.height(16.dp))
        Text(
            text = "Preview:",
            style = MaterialTheme.typography.labelMedium,
            fontWeight = FontWeight.Bold,
        )
        Spacer(modifier = Modifier.height(4.dp))
        val transposed = ChordSheetTranspose.transpose(content, transposeSemitones)
        Text(
            text = transposed,
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

@Composable
private fun StepReviewAndSave(
    songTitle: String,
    songContent: String,
    transposeSemitones: Int,
    selectedRoot: Int,
    chosenDegrees: List<ChordDegree>,
    selectedScale: ScaleType,
) {
    Text(
        text = "Review and save",
        style = MaterialTheme.typography.titleMedium,
        fontWeight = FontWeight.Bold,
        modifier = Modifier.semantics { heading() },
    )
    Spacer(modifier = Modifier.height(4.dp))
    Text(
        text = "Check your song below, then tap Save to add it to your songbook.",
        style = MaterialTheme.typography.bodyMedium,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
    Spacer(modifier = Modifier.height(16.dp))

    val keyName = Notes.pitchClassToName(selectedRoot)
    Text(
        text = "Key: $keyName ${selectedScale.label}",
        style = MaterialTheme.typography.labelMedium,
        fontWeight = FontWeight.Bold,
        color = MaterialTheme.colorScheme.primary,
    )

    if (chosenDegrees.isNotEmpty()) {
        val chordNames = chosenDegrees.map { degree ->
            val chordRoot = (selectedRoot + degree.interval) % Notes.PITCH_CLASS_COUNT
            Notes.enharmonicForKey(chordRoot, selectedRoot) + degree.quality
        }
        Text(
            text = "Progression: ${chordNames.joinToString(" – ")}",
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }

    Spacer(modifier = Modifier.height(8.dp))

    if (transposeSemitones != 0) {
        Text(
            text = "Transpose: ${ChordSheetTranspose.semitoneLabel(transposeSemitones)} semitones",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.tertiary,
        )
        Spacer(modifier = Modifier.height(4.dp))
    }

    Text(
        text = songTitle.ifBlank { "Untitled" },
        style = MaterialTheme.typography.titleSmall,
        fontWeight = FontWeight.Bold,
    )
    Spacer(modifier = Modifier.height(4.dp))

    val finalContent = if (transposeSemitones != 0) {
        ChordSheetTranspose.transpose(songContent, transposeSemitones)
    } else {
        songContent
    }
    Text(
        text = finalContent.ifEmpty { "(no lyrics yet)" },
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant,
    )
}
